keps/sig-api-machinery/20200123-client-go-ctx.md

```diff
@@ -0,0 +1,186 @@
+---
+title: Context support in k8s.io/client-go
+authors:
+- "@mikedanese"
+- "@maleck13"
+owning-sig: sig-api-machinery
+participating-sigs:
+- sig-api-machinery
+approvers:
+- "@deads2k"
+- "@liggitt"
+- "@lavalamp"
+creation-date: 2020-01-23
+last-updated: 2020-01-23
+status: provisional
+---
+
+# Context support in k8s.io/client-go
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This proposal outlines a refactoring that would allow propagating of
+[context](https://golang.org/pkg/context/) with
+[client-go](https://github.com/kubernetes/client-go) with API calls made using
+k8s.io/client-go clientsets.
+
+## Motivation
+
+When using client-go, external calls to the Kubernetes API are made in order to
+manage resources. Adding support for context propagation along these request
+paths enables clean and efficient implementation of cross cutting functionality.
+This functionality initially will include:
+
+- Support for request timeout and cancelation that frees the calling thread. Due
+  to issues with authentication and authorization performance, this was
+  something that was
+  [retrofited](https://github.com/kubernetes/kubernetes/pull/83064) in various
+  auth\* expansions but is generally useful.
+- Support for distributed tracing. Active
+  [efforts](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0034-distributed-tracing-kep.md#propagating-context-through-objects)
+  by SIG Instrumentation to support distributed tracing are either blocked or
+  would be substantially simplified with this enhancement.
+
+### Goals
+
+- Allow consumers of k8s.io/client-go to propagate context with API requests
+  made using a clientset.
+
+### Non-Goals
+
+- Plumb context to client-go callsites in core Kubernetes. We believe that by
+  making this feature available, this type of refactoring will happen over time.
+
+## Proposal
+
+The recommended approach is to pass the context with each action on a resource.
+We will modify the signatures of all client interfaces to accept a
+`context.Context` as the first argument.  This is idiomatic, explicit, and
+produces the least error prone API.
+
+## Design Details
+
+This is an example change to `Pods.Get` implementation:
+
+```go 
+func (c *pods) Get(ctx context.Context, name string, options meta_v1.GetOptions) (result *v1.Pod, err error) {
+	Result = &v1.Pod{}
+	err = c.client.Get().
+		Context(ctx).
```

Comment:
Interesting. I expected to see context passed into methods like Do and/or Into

Reply:
`Get` returns a new `Request` - the context belongs to that request?

Reply:
This [already exists](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request.go#L441-L446) in the Request type. Everything after Get/Create/Update is part of the request-specific object. Using the existing context setter on the request seems ok to me.

Reply:
> This [already exists](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request.go#L441-L446) in the Request type. Everything after Get/Create/Update is part of the request-specific object. Using the existing context setter on the request seems ok to me.

It makes it possible to have a mistake.  Other than, "it's here", I don't see why we would perpetuate code that is possible to mis-use if we didn't have to.

Reply:
The request object is already single-shot (it consumes the body reader, etc). Having a context member and a context parameter to `Do` is confusing (if both are specified, I don't think we can merge them)

Reply:
> The request object is already single-shot (it consumes the body reader, etc). Having a context member and a context parameter to `Do` is confusing (if both are specified, I don't think we can merge them)

The context where you decide to make a request is not necessarily the same as the context where you add the parameters or where you finally make the request.   This example conveniently has them all together, but it doesn't have to be that way.

I would remove the .Context() or at the very least deprecate it.

Reply:
We should try to avoid repeated changes of widely used public APIs so I think it's worth entertaining whether we should add Options in this batch. The set of callesites that would be affected by refactoring ctx handling in rest.Request is both much smaller and disjoint with the set of callsites that would be affected by the current proposed change. This seems like it could be easily sorted out in a refactor PR.

Reply:
> We should try to avoid repeated changes of widely used public APIs so I think it's worth entertaining whether we should add Options in this batch. 

I think we should. @deads2k agreed [here](https://github.com/kubernetes/enhancements/pull/1503#issuecomment-578892831)

> The set of callesites that would be affected by refactoring ctx handling in rest.Request is both much smaller and disjoint with the set of callsites that would be affected by the current proposed change. This seems like it could be easily sorted out in a refactor PR.

If it touches the Request#Do method, it'll affect all generated clients, right (including ones we intended to snapshot and not touch for two releases)?

Reply:
I don't think "snapshot and not touch" is the right characterization of these package. What we want to snapshot is the API, so that we have implementations of both the old and new APIs in tree for a 2 release window to allow incremental migration. For that to work, these packages need to get bug fixes, security fixes, and continue to compile.

Reply:
Yeah, sorry, that wasn't quite what I meant. If we address the request context plumbing first, we get the generator to do the work for us.

If picking up 1.18 means getting a signature change in Request#Do AND the code-generator, downstream consumers with their own generated packages don't have a released level of code-generator that will make their current client interface use the new request Do signature. 

I'm happy to defer the decision on the Request signature to the PR. The main thing I care about here is the client interface signatures.